controlpanel/api/rules.py

```diff
@@ -0,0 +1,194 @@
+"""
+Rules and permissions (and object permissions) using django-rules
+"""
+
+from rules import add_perm, is_active, is_authenticated, is_superuser, predicate
+
+from controlpanel.api.models import (
+    App,
+    AppS3Bucket,
+    Parameter,
+    S3Bucket,
+    UserApp,
+    UserS3Bucket,
+)
+from controlpanel.api.tools import ToolDeployment
+
+
+@predicate
+def is_app_admin(user, obj):
+    """
+    Check whether `user` is an admin for either;
+    - the `obj` `App`
+    - the `app` of the `obj` `AppS3Bucket`
+    - the `app` of the `obj` `UserApp`
+
+    :param user User: The user to check
+    :param obj App|AppS3Bucket|UserApp: The object to check
+    """
+    if obj is None:
+        return True
+
+    if isinstance(obj, App):
+        return user in obj.admins
+
+    if isinstance(obj, AppS3Bucket):
+        return is_app_admin(user, obj.app)
+
+    if isinstance(obj, UserApp):
+        return obj.is_admin
+
+    # XXX raise exception?
+    return False
+
+
+add_perm('api.list_app', is_authenticated)
+add_perm('api.create_app', is_authenticated & is_superuser)
```

Comment:
I imagine the permissions records associated with a superuser are deleted or updated if for whatever reason he stops being a superuser.

Reply:
This `is_superuser` predicate checks the `User.is_superuser` field. Django Rules doesn't use the `django.contrib.auth.models.Permission` class, it is not database backed.